### ISSUE TITLE
Some Feedback

### ISSUE THREAD

**Issue by avuserow** [soundart/perl6-tweetnacl#3]
Hi, I saw the note in the README requesting some feedback. I had some golang code that used these primitives to do SecretBox encryption/decryption so I wrote a bit of P6 code to see how it compared.

I was able to interoperate with the golang code fairly easily so it is great to know that it is compatible with other implementations.

Having written some P6 nativecall-based code in the past, here's some design suggestions based on that. My general philosophy is that I want to use nativecall-based modules to hide the details of C behind things that make more sense from P6. The same thought applies to ctypes in Python or XS in Perl 5 or any other bindings around C language libraries.

To this point, modules should avoid exporting the C functions. So in `SecretKey.pm6`, you would avoid marking `crypto_secretbox_open_int` as `is export`. If you do think that consumers of your modules may want to import and use these directly, consider putting them all in a `Raw.pm` file, [similar to how `GTK::Simple` does it](https://github.com/perl6/gtk-simple/blob/master/lib/GTK/Simple/Raw.pm6).

Along these lines, avoid requiring `CArray` and other NativeCall types as input parameters. I've found that functions taking `CArray` can be represented by the `Blob` type, since Blob also permits `Buf` and `utf8`, both are fairly common. You can take advantage of some other modules to help here, see the modules in the `NativeHelpers::` namespace:  http://modules.perl6.org/search/?q=nativehelpers

You may also consider using subtypes of Blob: `subset KeyData of Blob where *.elems == 32`. Then you can use this as a type parameter.

I would also suggest unifiying the return types of the two `encrypt` multi methods, since it seems weird to have them return different types depending on if you provide a nonce or not. If you do want different return types, I suggest document this.

Aside from this, I found the code reliable so far, so that's excellent. Thanks for putting this out here so other people can use it, I really appreciate it. 

Not sure if you are still actively interested in Perl 6 and/or this code, but thought I would pass along the feedback in hopes that it helps. Let me know if you want any more information or examples and I'll see what I can do.

**Comment by soundart** [comment]
Hi, Thank you for the improvement and style suggestions. I think I agree mostly. 

But I still do not know howto do my test approach when implementing them. 
My perl6 knowledge has unfortunately not improved much in the past years.

- too many exports / testing related:

The broad usage of export for functions like crypto_secretbox_open_int was required by the unit tests which I ported from the nacl library:

t/07-crypto_secret_box.t:

```
my Int $ret;
$ret = crypto_secretbox_int($c,$m,163,$nonce,$firstkey);
is $ret,0, "success return code";
is-deeply $c, $exp, "encrypt";
```

I am not sure anymore, but I think the strange multi-methods with or without nonce were test related too: with a random nonce it is hard to predict the resulting encrypted data and specify the expected data for a test.

I do not know if it is possible to provide some functions only to unit tests and not to the user of the module, but something like this would be nice. 

So basically I have no consumers which might need these "Raw" or low-level functions except my test code.

- style related / CArray usage:
 
I think I avoided high-level data types (Buf, utf8) for things like cryptographic keys intentionally and used CArrays to pass them around.

I tried to make the usage easier and used 'holder' classes like Keypair  and Ciphertext which contain the data in CArray attributes so that the user of the library would mostly not see these data types. 

The unimplemented idea is to add functionality to serialize these data structures for storage and/or transport to these 'holder' classes.

**Comment by soundart** [comment]
I just ran the module tests again with perl6 from debian buster / rakudo 2018.12-5 
and they fail there due to a bug which as fixed later https://github.com/rakudo/rakudo/issues/2681

**Comment by avuserow** [comment]
I haven't gotten much time to get back to this, so sorry about the delays.

I'm thinking maybe it makes sense to only export some things and hide the rest behind some export tags? I've never fully understood all the export options in Perl6 and I don't remember tons of it right now, but maybe putting some of it as something like `is export(:ALL)` and then you can do `use Crypt::TweetNacl::SecretKey :ALL` in the tests (or consumer modules). Having tests for the native code is a nice idea so hopefully you can still preserve that.

For the CArray versus Buf/whatever, I think I will need to write a bit more of my application to make a better suggestion here.

**Comment by soundart** [comment]
Hi thank you very much, I complety missed the option to selectivly export symbols.
https://docs.perl6.org/language/modules#Exporting_and_selective_importing

I think I will refactor the code and provide a ":testing" tag for the Exports. 

Of cource I am interested  in your use case is for this module...
